[16.0] stock_release_channel: Manage channel lifecycle

lmignon opened the pull request:
* [x] includes #493
* [x] depends on #478 

The availability of a release channel depends on its state. A release channel
can be in one of the following states:

  - Open: the channel is available and can be used to release transfers. New
    transfer are assigned automatically to this channel.
  - Locked: the channel is available but the release of transfers from the channel
    is no more possible. New transfers are still automatically assigned to this
  - Asleep: the channel is not available and cannot be used to release
    transfers. It is also no more possible to assign transfers to this channel.

New release channels are by default "Open". You can change its state by using
the "Lock" and "Sleep" buttons. When the "Sleep" button is used, in addition to
the state change, not completed transfers assigned to the channel are unassigned
from the channel. When the "Lock" button is used, only the state change is changed.
A locked channel can be unlocked by using the "Unlock" button.
A asleep channel can be waken up by using the "Wake up" button. When the "Wake up"
button is used, in addition to the state change, the system looks for pending
transfers requiring a release and try to assign them to a channel in the
"Open" or "Locked" state.

lmignon committed:
[IMP] stock_release_channel: Makes assign_release_channel takes only 1 picking as args

Except from the testsuite, the assign_release_channel method from stock.release.channel model is always called with one picking as args. To allows future improvements add a specific contraint on this actual usage

lmignon committed:
[IMP] stock_release_channel: Allows to find the channel from a picking instance

lmignon committed:
[IMP] stock_release_channel: Allows to filter channel candidates into the assign_release_channel method

In some cases, you could need to prevent a channel to be used as candidate into the assign_release_channel mehtod. This change add a hook method that can be used to filter the channels

lmignon committed:
[IMP] stock_release_channel: Manage release channel lifecycle

The availability of a release channel depends on its state. A release channel
can be in one of the following states:

  - Open: the channel is available and can be used to release transfers. New
    transfer are assigned automatically to this channel.
  - Locked: the channel is available but the release of transfers from the channel
    is no more possible. New transfers are still automatically assigned to this
  - Asleep: the channel is not available and cannot be used to release
    transfers. It is also no more possible to assign transfers to this channel.

New release channels are by default "Open". You can change its state by using
the "Lock" and "Sleep" buttons. When the "Sleep" button is used, in addition to
the state change, not completed transfers assigned to the channel are unassigned
from the channel. When the "Lock" button is used, only the state change is changed.
A locked channel can be unlocked by using the "Unlock" button.
A asleep channel can be waken up by using the "Wake up" button. When the "Wake up"
button is used, in addition to the state change, the system looks for pending
transfers requiring a release and try to assign them to a channel in the
"Open" or "Locked" state.

jbaudoux commented on stock_release_channel/models/stock_release_channel.py:
```diff
@@ -621,3 +705,72 @@ def release_next_batch(self):
+        pickings_to_unassign = self.env["stock.picking"].search(
+            self._get_picking_to_unassign_domain()
+        )
+        pickings_to_unassign.write({"release_channel_id": False})
```
Before unassigning, the picking must be "unreleased".
Could be some ongoing pickings cannot be unreleased and this action will be declined

lmignon replied:
This PR doesn't depend on #478 I'll implement this requirement in #508

lmignon replied:
@jbaudoux done

lmignon committed:
[FIX] stock_release_channel: A released picking is a picking with a last_release_date

Before this change a released picking was a picking that do no more need to be released. With the management of backorders introduced in https://github.com/OCA/wms/pull/478, some released pickings could still need to be released since qties were not available for some product at release time. Fortunately, at the same time as this change in the management of the release process, a new field has been added to log the timestamp of the last release done on a picking. Therefore, to kown if picking is released we can replace the search on 'need_release' by a search on 'last_release_date'. A released picking is a picking with a last_release_date filled.

lmignon commented:
@simahawk @jbaudoux rebase done but  requires #478

rousseldenis reviewed (approved):
Code review.

@lmignon Could you remove the test-requirements commit ?

lmignon commented:
@rousseldenis test-requirements removed

jbaudoux reviewed (approved):
LG

jbaudoux commented:
/ocabot merge major
